tests/test_optimizers.py

```diff
@@ -215,3 +223,164 @@ class TestFtrlOptimizer(GPflowTestCase, OptimizerCase):
 # E               TypeError: Input 'lr' of 'ApplyProximalAdagrad' Op has type float32 that does not match type float64 of argument 'var'.
 #
 # ../../anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py:546: TypeError
+
+def test_VGP_vs_GPR(session_tf):
+    N, D = 3, 2
+    X = np.random.randn(N, D)
+    Y = np.random.randn(N, 1)
+    kern = gpflow.kernels.RBF(D)
+    lik_var = 0.1
+    lik = gpflow.likelihoods.Gaussian()
+    lik.variance = lik_var
+
+    m_vgp = gpflow.models.VGP(X, Y, kern, lik)
+    m_gpr = gpflow.models.GPR(X, Y, kern)
+    m_gpr.likelihood.variance = lik_var
+
+    m_vgp.set_trainable(False)
+    m_vgp.q_mu.set_trainable(True)
+    m_vgp.q_sqrt.set_trainable(True)
+    NatGradOptimizer(1.).minimize(m_vgp, [[m_vgp.q_mu, m_vgp.q_sqrt]], maxiter=1)
+
+    assert_allclose(m_gpr.compute_log_likelihood(),
+                    m_vgp.compute_log_likelihood(), atol=1e-5)
+
+def test_SVGP_vs_SGPR(session_tf):
+    N, M, D = 4, 3, 2
+    X = np.random.randn(N, D)
+    Z = np.random.randn(M, D)
+    Y = np.random.randn(N, 1)
+    kern = gpflow.kernels.RBF(D)
+    lik_var = 0.1
+    lik = gpflow.likelihoods.Gaussian()
+    lik.variance = lik_var
+
+    m_svgp = gpflow.models.SVGP(X, Y, kern, lik, Z=Z)
+    m_sgpr = gpflow.models.SGPR(X, Y, kern, Z=Z)
+    m_sgpr.likelihood.variance = lik_var
+
+    m_svgp.set_trainable(False)
+    m_svgp.q_mu.set_trainable(True)
+    m_svgp.q_sqrt.set_trainable(True)
+    NatGradOptimizer(1.).minimize(m_svgp, [[m_svgp.q_mu, m_svgp.q_sqrt]], maxiter=1)
+
+    assert_allclose(m_sgpr.compute_log_likelihood(),
+                    m_svgp.compute_log_likelihood(), atol=1e-5)
+
+
+class CombinationOptimizer(Optimizer):
+    """
+    A class that applies one step of each of multiple optimizers in a loop
+    """
+    def __init__(self, optimizers_with_kwargs):
+        self.name = None
+        super().__init__()
+        self.optimizers_with_kwargs = optimizers_with_kwargs
+
+    def minimize(self, model, session=None, maxiter=1000, feed_dict=None, anchor=True):
+        session = model.enquire_session(session)
+
+        minimize_ops = []
+        for optimizer, kwargs in self.optimizers_with_kwargs:
+            # hack to init the optimizer operation
+            optimizer.minimize(model, maxiter=0, **kwargs)
+
+            # this is what we will call in the loop
+            minimize_ops.append(optimizer.minimize_operation)
+
+        with session.graph.as_default(), tf.name_scope(self.name):
+            feed_dict = self._gen_feed_dict(model, feed_dict)
+
+            for _i in range(maxiter):
+                for op in minimize_ops:
+                    session.run(op, feed_dict=feed_dict)
+
+        if anchor:
+            model.anchor(session)
+
+# # TODO(@hugh) this needs to be fixed
```

Comment:
What's wrong with this code? We should fix it before merging.

Reply:
This will be superseded by the Actions PR

Reply:
The problem is that if the q_mu, q_sqrt params are not trainable their change doesn't get noticed by the ordinary grad optimizer

Reply:
Oh no!!! I've just realized what's wrong with this case!